### ISSUE TITLE
Out-of-memory error during conversion from AnnData to Seurat if X is numpy array.

### ISSUE THREAD

**Issue by pablo-gar** [chanzuckerberg/single-cell-curation#33]
TLDR: An upload to the corpora data portal can fail due to an out-of-memory error during the conversion from AnnData to Seurat when X is a numpy array. Currently, there's no error message that indicates so. This could be avoided if X is always a scipy sparse matrix.

The issue was discovered while processing this:

- [Dataset](https://cellxgene.cziscience.com/e/d33814fc-31e8-4f39-a200-a0ca21a9b134.cxg/)
- [Collection](https://cellxgene.cziscience.com/collections/5e469121-c203-4775-962d-dcf2e5d6a472)

And the process of discovery was the following

1. The dataset upload failed, however the explorer was working ok. The downloads were available for the AnnData and Loom files but not the Seurat file -- indicating that conversion to Seurat was causing the issue. The error was raised when calling the [update_db()](https://github.com/chanzuckerberg/corpora-data-portal/blob/ffca067b9e4aea237fa2bd7c7a9cbc5813ebd449/backend/corpora/dataset_processing/process.py#L225) function from [process_cxg() ](https://github.com/chanzuckerberg/corpora-data-portal/blob/ffca067b9e4aea237fa2bd7c7a9cbc5813ebd449/backend/corpora/dataset_processing/process.py#L392)during the processing of the upload.
2. Trying to manually run the conversion in a 700+G machine gave this error from R `'Realloc' could not re-allocate memory (18446744070851829760 bytes)`.
3. Upon closer inspection, the count matrix was a numpy array. This leads to R trying to load the entire matrix inefficiently and ends up running out of memory. Changing it to a scipy sparse matrix solves the issue as R loads it as sparse object as well.
4. Processing in the portal works after shifting to scipy sparse matrix.

There are two potential solutions I thought of:

- We require users to always use scipy sparse matrices when creating the h5ad file (and implement this as well in cellxgene-schema apply). Then the validator should check for this requirement
- We transform numpy arrays to scipy sparse matrices if any are found during the processing of uploads. 

It's worth noting that while these solutions will likely reduce the number of times we run into out-of-memory issues, they don't guarantee the issues won't happen again. Moreover there's no error message delivered when this happens, neither in the backend or frontend.

**Comment by brianraymor** [comment]
Reminder that we're waiting on a related [proposal](https://chanzuckerbergteam.slack.com/archives/C01CPJURT0S/p1616191273069500?thread_ts=1616019680.035500&cid=C01CPJURT0S) from engineering for _Add logging that helps us more clearly identify out-of-memory errors_.

**Comment by maniarathi** [comment]
Created #1013 and #1014

**Comment by pablo-gar** [comment]
Bruce suggested:

> - H5AD's that are mathematically sparse (cutoff TBD, but certainly >50%) should be saved as a CSR, with the compress="gzip" option
> - This should be a curator policy decision, not hard-wired into the Portal
> - Perhaps it would be good to have a "lint" tool that curators could use before submitting.  Ie, check for format, schema, etc -- all optional "suggestions" about appropriate policy.

**Comment by ambrosejcarr** [comment]
Bruce's suggestion that we require sparse data files be submitted using sparse formats makes sense to me. I can update the encoding to specify this requirement over X if there is agreement within this group. 

I agree that the portal needn't know about this requirement.

**Comment by brianraymor** [comment]
@ambrosejcarr and @pablo-gar - suggest we close this issue. `cellxgene-schema` will warn for this case - but is not performing automatic conversion. 

> In any layer, if a matrix has 50% or more values that are zeros, it is STRONGLY RECOMMENDED that the matrix be encoded as a scipy.sparse.csr_matrix.

**Comment by pablo-gar** [comment]
Agreed, it can be closed.
